README.md
# Design of evolved synthetic 5'UTR sequences for enhanced protein production

The goal of this workflow is to generate novel 5'UTR sequences with optimized features to maximize the translation efficiency of the constructs. 
This is achieved in 4 steps:

- Step 1. **Feature extraction** from endogenous sequences 
- Step 2. **Model generation** using extracted features to predict Translation Efficiency (TE)
- Step 3. **Model evaluation** 
- Step 4. **Generation of novel 'evolved' sequences** using a Genetic Algorithm

This repository is the code accompanying the paper:High-Throughput 5’ UTR Engineering for Enhanced Protein Production in Non-Viral Gene Therapies. Jicong Cao*, Eva Maria Novoa*, Zhizhuo Zhang*, William C.W. Chen, Dianbo Liu, Gigi C G Choi, Alan S L Wong, Claudia Wehrspaun, Manolis Kellis, Timothy K Lu. bioRxiv 2020. doi: https://doi.org/10.1101/2020.03.24.006486


![alt text](./img/init_fig.png "init_fig")


## Table of Contents
- [Step 1. Feature extraction of 5'UTR sequences](#step-1-feature-extraction-of-5utr-sequences)
- [Step 2. 5'UTR model generation](#step-2-5utr-model-generation)
- [Step 3. 5'UTR model evaluation](#step-3-5utr-model-evaluation)
- [Step 4. Generation of novel evolved 5'UTR sequences](#step-4-generation-of-novel-evolved-5utr-sequences)
- [Additional data sources](#Additional-data-sources)
- [Dependencies and versions](#Dependencies-and-versions)
- [Citation](#Citation) 
- [Contact](#Contact) 
 

## Step 1. Feature extraction of 5'UTR sequences 

**Goal**: Extract features that correspond to optimal 5'UTRs to maximize the translation efficiency (TE).

The features extracted from endogenous sequences (which will be used to generate models later) are:
- K-mer frequency: k=1-6
- RNA folding energy for: i) the first 100bp of the 5'UTR, ii) last 30bp(15bp UTR+ 15bp CDS), iii) whole (5UTR+15bp CDS), iv) 5'UTR only, with/without considering G-quadruplex 
- Codon usage (of the 15 first bp of the CDS region)
- Number of start and stop codons in teh 5'UTR
- 5'UTR length

### Running the code: 
``` 
python run_pipeline.py feature_extract --input_fasta data/gencode_v17_5utr_15bpcds.fa --output_dir output/
```

## Step 2. 5'UTR model generation

**Goal**: Generate model  that correspond to optimal 5'UTRs to maximize the translation efficiency (TE).

**Algorithm used**: **Random Forest** (1) to build a prediction model is recommended, however the user can also choose to run glmnet, Rpart and SVM. 

The model trained for human sequences is available as part of this repository. 

### Running the code: 
``` 
python run_pipeline.py model_build --prefix output/input.fa --annotation_file data/df_counts_and_len.TE_sorted.Muscle.with_annot.txt --min_rna_rpkm 5 --min_riboseq_rpkm 0.1 --model 1 --out output/muscle_randomforest.model
```

## Step 3. 5'UTR model evaluation

**Goal**: Evaluate different type of machine learning models in predicting translation efficiency, and decide which model to use in the next step (models are evaluated in terms of Spearman correlation and root mean square error). 

Results: Using 10-fold cross-validation, we obtained  0.71 Spearman correlation in TE prediction, and 0.74 in RNA expression prediction.


#### Running the code:

```
python run_pipeline.py model_eval --prefix output/input.fa --annotation_file data/df_counts_and_len.TE_sorted.Muscle.with_annot.txt --min_rna_rpkm 5 --min_riboseq_rpkm 0.1 --modellist 1,2,3,4 --out output/muscle_eval.xlsx
```


## Step 4. Generation of novel evolved 5'UTR sequences

**Goal**: Evolve endogenous 5'UTR sequences to obtain 5'UTRs with increased translation efficiency

**Algorithm used**: **Genetic Algorithm (GA)**

Additional details of individual steps are described in `makefile` file)

#### Running the code:
```
python run_pipeline.py  sequence_generate --n_total 3585 --prefix output/input.fa --annotation_file data/df_counts_and_len.TE_sorted.Muscle.with_annot.txt --min_rna_rpkm 5 --min_riboseq_rpkm 0.1 -t ribo  -m output/muscle_randomforest.model  -o output/
```

## Additional data sources 

- df_counts_and_len.TE_sorted.with_annot.txt: Muscle RNA-seq and Ribo-seq data from publicly available datasets (PC3, HEK and muscle)
- entiredata_2015_12_25_3-25_pm: RNA binding motif, downloaded from http://cisbp-rna.ccbr.utoronto.ca/
 
## Dependencies and versions

#### Create a Conda Environment with all the dependencies:
```
conda env create -f environment.yml
```

Here we list the specific versions of individual dependencies needed:

| Software | Version |
| ------------- | ------------- |
| python  | 3.5  |
| viennarna  | 2.1.9  |
| R  | 3.5.1  |
| biopython  | 1.72 |
| r-randomforest  | 4.6.14 |
| r-ga  | 3.2  |
| r-seqinr  | 3.6.1  |
| r-glmnet  | 2.0.16  |


## Citation
If you find this work useful, please cite: 

High-Throughput 5’ UTR Engineering for Enhanced Protein Production in Non-Viral Gene Therapies. Jicong Cao*, Eva Maria Novoa*, Zhizhuo Zhang*, William C.W. Chen, Dianbo Liu, Gigi C G Choi, Alan S L Wong, Claudia Wehrspaun, Manolis Kellis, Timothy K Lu. bioRxiv 2020. doi: https://doi.org/10.1101/2020.03.24.006486


## Contact
If you have any issues using this code, please open an Issue in the GitHub repository. Thanks!


#### Last Updated: February 2021



changelog.txt
1. change repo name to 5UTR_Optimizer
2. add parameter options and restructure buildModel_final.R
3. create new R script evalModel_final.R for evaluation different choice of ML model
4. rename evolutionDesign_Ribo.R to evolutionDesign.R, and add corresponding parameters options
5. modify FeatureCommons.py, FeatureExtraction_final.py, FeatureExtraction_singleInput.py, finalFormat_3k_synthetic_seqs.py to be compatible for python 3
6. add environment.yml for all dependencies
7. modified readme with updated command example and software version
8. deleted unneccessary old scripts

Feb 27 2021
1. implement glmnet/rpart/svm and generate plot pdf for spearcor, R-square/normliazed version, feature importance
2. update readme to put evalmodel before the build model
3. python script plot_eval.py to plot excel file data generated by evalModel_final.R
4. output excel contains six tabs:
    - TE_result: the evaluation result for predicing TE for each model in 10 fold CVs  , sp_cor is spearman correlation, rsq is R-squared statstics
    - RNA_result:  the evaluation result for predicing RNA-seq RPKM for each model in 10 fold CVs  , sp_cor is spearman correlation, rsq is R-squared statstics
    - TE_features_accuracy: random-forest feature importance value based on prediction accuracy for TE
    - TE_features_impurity: random-forest feature importance value based on node impurity for TE
    - RNA_features_accuracy: random-forest feature importance value based on prediction accuracy for RNA-seq RPKM
    - RNA_features_impurity: random-forest feature importance value based on node impurity for RNA-seq RPKM

run_pipeline.py
import sys,os
import argparse
def get_cmd_parser():
    main_parser = argparse.ArgumentParser(add_help=False)

    main_parser.add_argument('-v', '--version', action='version',
                        version='%(prog)s 1.0', help="Show program's version number and exit.")
    main_parser.add_argument('-h', '--help', action='help', default=argparse.SUPPRESS,
                        help='run_pipeline.py [task]  [task parameters]',)



    subparsers = main_parser.add_subparsers(help='feature_extract | model_build | model_eval | sequence_generate | full_pipe', dest='taskname')
    feature_extraction_parser = subparsers.add_parser(name="feature_extract",
                                                    help="""feature extraction step, convert fasta file to sequence feature format.
                                                     Example: feature_extract --input_fasta data/gencode_v17_5utr_15bpcds.fa --output_dir output/""")

    feature_extraction_parser.add_argument("--input_fasta",required=True,help="5utr sequences in fasta format")

    feature_extraction_parser.add_argument('--output_dir',
                        default='./output',
                        help='the output directory')


    model_build_parser = subparsers.add_parser(name="model_build",
                                                    help="""build machine learning model from given sequence features and label data 
                                                    Example: model_build --prefix output/input.fa --annotation_file data/df_counts_and_len.TE_sorted.Muscle.with_annot.txt --min_rna_rpkm 5 --min_riboseq_rpkm 0.1 --model 1 --out output/muscle_randomforest.model
                                                    """)

    model_build_parser.add_argument("-p", "--prefix",required=True,help="sequence feature files prefix")
    model_build_parser.add_argument("-k", "--min_rna_rpkm", help="minimal RNASEQ RPKM for the gene included in training. In paper, we used 5 for muscle cell")
    model_build_parser.add_argument("-r", "--min_riboseq_rpkm", help="minimal RiboSEQ RPKM for the gene included in training. In paper, we used 0.1 for muscle cell")
    model_build_parser.add_argument("-a", "--annotation_file", required=True,help="transcript annotation file with gene id and RPKM information for rnaseq and riboseq")
    model_build_parser.add_argument("-m", "--model" , default=1, help="1: randomforest, 2: glmnet, 3: regression tree, 4: SVM  [default= %default]")
    model_build_parser.add_argument("-j", "--n_cpus", default=16, help="number of cpus [default= %default]")
    model_build_parser.add_argument("-o", "--out", default="trained.model", help="output file name [default= %default]")




    model_eval_parser = subparsers.add_parser(name="model_eval",
                                                    help="eval the built ML model using cross validation")


    model_eval_parser.add_argument("-p", "--prefix",required=True,help="sequence feature files prefix")
    model_eval_parser.add_argument("-k", "--min_rna_rpkm",default=50, help="minimal RNASEQ RPKM for the gene included in training. In paper, we used 5 for muscle cell")
    model_eval_parser.add_argument("-r", "--min_riboseq_rpkm", default=5,help="minimal RiboSEQ RPKM for the gene included in training. In paper, we used 0.1 for muscle cell")
    model_eval_parser.add_argument("-a", "--annotation_file", required=True,help="transcript annotation file with gene id and RPKM information for rnaseq and riboseq")
    model_eval_parser.add_argument("-m", "--modellist" , default="1,2,3,4", help="1: randomforest, 2: glmnet, 3: regression tree, 4: SVM")
    model_eval_parser.add_argument("-j", "--n_cpus", default=16, help="number of cpus [default= %default]")
    model_eval_parser.add_argument("-o", "--out", default="evaluation.pdf", help="output evaluation pdf file name [default= %default]")

    sequence_generate_parser = subparsers.add_parser(name="sequence_generate",
                                                    help="""generate synthetic UTR sequences with optimized Riob level or transcription efficiency based on the pretraned ML model\
                                                    Example: sequence_generate --prefix output/input.fa --annotation_file data/df_counts_and_len.TE_sorted.Muscle.with_annot.txt --min_rna_rpkm 5 --min_riboseq_rpkm 0.1 -t ribo  -m output/muscle_randomforest.model  -o output/
                                                    """)

    sequence_generate_parser.add_argument("-p", "--prefix",required=True,help="sequence feature files prefix")
    sequence_generate_parser.add_argument("-k", "--min_rna_rpkm",default=50, help="minimal RNASEQ RPKM for the gene included in training. In paper, we used 5 for muscle cell")
    sequence_generate_parser.add_argument("-r", "--min_riboseq_rpkm",default=5, help="minimal RiboSEQ RPKM for the gene included in training. In paper, we used 0.1 for muscle cell")
    sequence_generate_parser.add_argument("-a", "--annotation_file", required=True,help="transcript annotation file with gene id and RPKM information for rnaseq and riboseq")
    sequence_generate_parser.add_argument("-m", "--model_file" ,required=True,   help="path to the saved model file")
    sequence_generate_parser.add_argument("-j", "--n_cpus", default=16, help="number of cpus [default= %default]")
    sequence_generate_parser.add_argument("-o", "--out", default="./output", help="output folder path [default= %default]")
    sequence_generate_parser.add_argument("-t", "--optimize_target", default="ribo",
                                          help="optimize_target: ribo | te  [default= %default]")
    sequence_generate_parser.add_argument("-n", "--n_total", default=3585,
                                          help="total number of synthetic sequences")

    ##### full pipeline section ######
    fullpipe_parser = subparsers.add_parser(name="full_pipe",
                                                    help="run full pipeline from start to the end: feature_extract -> model_build -> model_eval -> sequence_generate ")

    fullpipe_parser.add_argument("-i","--input_fasta", required=True, help="5utr sequences in fasta format")

    fullpipe_parser.add_argument("-o",'--output_dir',
                        default='./output',
                        help='the output directory')

    fullpipe_parser.add_argument("-k", "--min_rna_rpkm", default=50,help="minimal RNASEQ RPKM for the gene included in training. In paper, we used 5 for muscle cell")
    fullpipe_parser.add_argument("-r", "--min_riboseq_rpkm", default=5,help="minimal RiboSEQ RPKM for the gene included in training. In paper, we used 0.1 for muscle cell")
    fullpipe_parser.add_argument("-a", "--annotation_file", help="transcript annotation file with gene id and RPKM information for rnaseq and riboseq")
    fullpipe_parser.add_argument("-m", "--model" , default=1, help="1: randomforest, 2: glmnet, 3: regression tree, 4: SVM  [default= %default]")
    fullpipe_parser.add_argument("-j", "--n_cpus", default=16, help="number of cpus [default= %default]")
    fullpipe_parser.add_argument("-n", "--n_total", default=3585,
                                          help="total number of synthetic sequences")



    return main_parser

def oss(cmd):
    print("executing:", cmd)
    os.system(cmd)

if __name__ == '__main__':
    parser=get_cmd_parser()
    args = parser.parse_args()
    if args.taskname=="feature_extract" or args.taskname=="full_pipe":
        os.makedirs(args.output_dir,exist_ok=True)
        cmd="python FeatureExtraction_final.py %s %s"%(args.input_fasta,args.output_dir)
        oss(cmd)

        if args.taskname=="full_pipe":
            args.prefix=args.output_dir+"/input.fa"



    if args.taskname=="model_build" or args.taskname=="full_pipe":
        if args.taskname=="full_pipe":
            args.out=args.output_dir+"/trained.model"
        cmd="Rscript buildModel_final.R -p %s -a %s -k %s -r %s -m %s -o %s"%(args.prefix,
                                                                              args.annotation_file,
                                                                            args.min_rna_rpkm,
                                                                             args.min_riboseq_rpkm,
                                                                              args.model,args.out  )
        oss(cmd)

    if args.taskname=="model_eval" :
        cmd="Rscript evalModel_final.R -p %s -a %s -k %s -r %s -m %s -o %s"%(args.prefix,
                                                                              args.annotation_file,
                                                                            args.min_rna_rpkm,
                                                                             args.min_riboseq_rpkm,
                                                                              args.modellist,args.out  )
        oss(cmd)

    if args.taskname=="sequence_generate" or args.taskname=="full_pipe":
        for seed in range(int(args.n_total)):
            if args.taskname=="full_pipe":
                args.model_file=args.output_dir+"/trained.model"
                for optimize_target in ['ribo','te']:
                    cmd = "Rscript evolutionDesign.R -p %s -a %s -t %s -k %s -r %s -m %s  -o %s" % (args.prefix,
                                                                                                args.annotation_file,
                                                                                                optimize_target,
                                                                                                args.min_rna_rpkm,
                                                                                                args.min_riboseq_rpkm,
                                                                                                args.model_file, args.out)
                    oss(cmd)

            else:
                cmd="Rscript evolutionDesign.R -p %s -a %s -t %s -k %s -r %s -m %s  -o %s"%(args.prefix,
                                                                                      args.annotation_file, args.optimize_target,
                                                                                    args.min_rna_rpkm,
                                                                                     args.min_riboseq_rpkm,
                                                                                      args.model_file,args.out  )
                oss(cmd)
